src/node_file.cc

```diff
@@ -551,51 +551,139 @@ void AfterScanDir(uv_fs_t* req) {
+        return;
       }
+      name_idx = 0;
+    }
+  }
 
-      MaybeLocal<Value> filename =
-          StringBytes::Encode(env->isolate(),
-                              ent.name,
-                              req_wrap->encoding(),
-                              &error);
-      if (filename.IsEmpty())
-        return req_wrap->Reject(error);
+  if (name_idx > 0) {
+    fn->Call(env->context(), names, name_idx, name_argv)
+      .ToLocalChecked();
+  }
 
-      name_argv[name_idx++] = filename.ToLocalChecked();
+  req_wrap->Resolve(names);
+}
 
-      if (name_idx >= arraysize(name_argv)) {
-        fn->Call(env->context(), names, name_idx, name_argv)
-            .ToLocalChecked();
-        name_idx = 0;
+void AfterScanDirWithTypes(uv_fs_t* req) {
+  FSReqBase* req_wrap = FSReqBase::from_req(req);
+  FSReqAfterScope after(req_wrap, req);
+
+  if (!after.Proceed()) {
+    return;
+  }
+
+  Environment* env = req_wrap->env();
+  Local<Value> error;
+  int r;
+  Local<Array> names = Array::New(env->isolate(), 0);
+  Local<Function> fn = env->push_values_to_array_function();
+  Local<Value> name_argv[NODE_PUSH_VAL_TO_ARRAY_MAX];
+  size_t name_idx = 0;
+  Local<Value> types = Array::New(env->isolate(), 0);
+  Local<Value> type_argv[NODE_PUSH_VAL_TO_ARRAY_MAX];
+  size_t type_idx = 0;
+
+  for (int i = 0; ; i++) {
+    uv_dirent_t ent;
+
+    r = uv_fs_scandir_next(req, &ent);
+    if (r == UV_EOF)
+      break;
+    if (r != 0) {
+      return req_wrap->Reject(
+          UVException(r, nullptr, req_wrap->syscall(),
+            static_cast<const char*>(req->path)));
+    }
+
+    MaybeLocal<Value> filename =
+      StringBytes::Encode(env->isolate(),
+          ent.name,
+          req_wrap->encoding(),
+          &error);
+    if (filename.IsEmpty())
+      return req_wrap->Reject(error);
+
+    name_argv[name_idx++] = filename.ToLocalChecked();
+
+    if (name_idx >= arraysize(name_argv)) {
+      MaybeLocal<Value> ret = fn->Call(env->context(), names, name_idx,
+                                       name_argv);
+      if (ret.IsEmpty()) {
+        return;
       }
+      name_idx = 0;
     }
 
-    if (name_idx > 0) {
-      fn->Call(env->context(), names, name_idx, name_argv)
-          .ToLocalChecked();
+    type_argv[type_idx++] = Integer::New(env->isolate(), ent.type);
+
+    if (type_idx >= arraysize(type_argv)) {
+      MaybeLocal<Value> ret = fn->Call(env->context(), types, type_idx,
+          type_argv);
+      if (ret.IsEmpty()) {
+        return;
+      }
+      type_idx = 0;
+    }
+  }
+
+  if (name_idx > 0) {
+    MaybeLocal<Value> ret = fn->Call(env->context(), names, name_idx,
+        name_argv);
+    if (ret.IsEmpty()) {
+      return;
     }
+  }
 
-    req_wrap->Resolve(names);
+  if (type_idx > 0) {
+    MaybeLocal<Value> ret = fn->Call(env->context(), types, type_idx,
+        type_argv);
+    if (ret.IsEmpty()) {
+      return;
+    }
   }
+
+  Local<Array> result = Array::New(env->isolate(), 2);
```

Comment:
Perhaps keeping a local constant array in JS, which is passed to the C++ layer, can be reused instead of creating a new array of length 2 every time? Just be sure to empty out the contents afterwards to allow GC.

Reply:
Hmm, I'm not too sure about the mechanics of this, due to the async version. Would stashing the array on the `FSReqBase` be the thing to do here? Or in the node environment (and then only passing it in once)?

Reply:
There would be no problem for async since setting the two array elements and processing them would happen on the main thread and would happen at the same time.

As far as how to allocate the array only once, you could do it any which way really. I don't really have a preference.

Reply:
I wouldn’t expect all too much overhead from creating a new array in C++ every time, so maybe we can leave it at this…?

Something where we probably can save time is using a typed array for the second entry, instead of a plain array of integers.

Reply:
Last I checked creating arrays or objects in C++ was significantly slower. Besides I would imagine having less pressure on the GC would be a good thing?

Reply:
It is going to be faster in JS, yes. I just don’t think it’s going to have an effect in the big picture, when compared to all the other work this code does, and I’m not a fan of optimizations like these when we haven’t measured their impact as part of their actual use case…

Reply:
@mscdex @addaleax Maybe the thing to do then, is punt on this for now, and do an optimization PR later after seeing if it turns out to be significant?